app/services/find_by_mime_type.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+# @abstract Returns or yields a set of FileResource objects based on their mime types.
+#
+# Examples
+#
+# Return a set of file resources for a given mime type
+#
+#   resources = FindByMimeType.call(mime_types: 'image/png')
+#
+# Return a set of file resources for a set of mime types
+#
+#   resources = FindByMimeType.call(mime_types: ['image/png', 'image/jpeg'])
+#
+# Iterate over all text-based file resources. This is helpful when needing to batch process a large set of files
+#
+#   FindByMimeType.call(mime_types: :text) do |resource|
+#     resource.update(...)
+#   end
+#
+
+class FindByMimeType
+  TEXT = %w(
+    application/msword
+    application/pdf
+    application/postscript
+    application/rtf
+    application/vnd.ms-excel
+    application/vnd.oasis.opendocument.spreadsheet
+    application/vnd.oasis.opendocument.text
+    application/vnd.openxmlformats-officedocument.presentationml.presentation
+    application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
+    application/vnd.openxmlformats-officedocument.wordprocessingml.document
+    application/vnd.wordperfect
+    application/x-mobipocket-ebook
+    html
+    text
+    xml
+  ).freeze
+
+  class << self
+    def call(mime_types: [])
+      types = mime_types == :text ? TEXT : Array.wrap(mime_types)
+      sql = build_statement(types)
+
+      ActiveRecord::Base
+        .connection
+        .execute(sql)
+        .values
+        .flatten
+        .map do |id|
+          FileResource.find(id) do |resource|
+            block_given? ? yield(resource) : resource
+          end
+        end
+    end
+
+    private
+
+      def build_statement(types)
+        inner_sql = types.map do |type|
+          "file_data->'metadata'->>'mime_type' LIKE '%#{type}%'"
+        end.join("\n            OR    ")
+
+        %(
+        SELECT DISTINCT
+          file_version_memberships.file_resource_id
+        FROM
+          file_version_memberships
+        WHERE
+          file_version_memberships.work_version_id IN (
+            SELECT id
+            FROM work_versions
+            WHERE aasm_state = 'published'
+          )
+        AND
+          file_version_memberships.file_resource_id IN (
+            SELECT id
+            FROM file_resources
+            WHERE #{inner_sql}
+          )
+        )
```

Comment:
Not sure how much of a performance impact this would have, but a join might be faster than a subquery. Might be worth testing it out to see if there is any meaningful difference. Don't let it hold this up, though. Just a suggestion.

Caution: untested! ☠️ ☠️ ☠️ 

```
SELECT DISTINCT
	file_version_memberships.file_resource_id
FROM
	file_version_memberships
JOIN 
	work_versions 
ON 
	file_version_memberships.work_version_id = work_versions.id
JOIN 
	file_resources 
ON 
	file_version_memberships.file_resource_id = file_resources.id
WHERE 
	work_versions.aasm_state = 'published'
AND 
	#{inner_sql}
```

Reply:
@nicktiberi Good idea! I tried that just now in the postgres shell, and the join query is ~10 ms _slower_ than my subquery approach. I wasn't expecting that because I would assume joins would be faster. My guess is that the join is operating on all the rows, whereas the subqueries reduce the number of rows on each table a bit. 🤷

Reply:
@awead - interesting! Yeah, I would have expected the join to be faster, but your theory makes sense. Subqueries are easier to read, anyway, IMO.

Reply:
FWIW, I did:

``` sql
SELECT DISTINCT
        file_version_memberships.file_resource_id
FROM
        file_version_memberships
JOIN
        work_versions
ON
        file_version_memberships.work_version_id = work_versions.id
JOIN
        file_resources
ON
        file_version_memberships.file_resource_id = file_resources.id
WHERE
        work_versions.aasm_state = 'published'
AND
        file_data->'metadata'->>'mime_type' LIKE '%pdf%'
OR
        file_data->'metadata'->>'mime_type' LIKE '%text%'
```

And got `Time: 49.560 ms`

Whereas 

``` sql
SELECT DISTINCT
  file_version_memberships.file_resource_id
FROM
  file_version_memberships
WHERE
  file_version_memberships.work_version_id IN (
    SELECT id
    FROM work_versions
    WHERE aasm_state = 'published'
  )
AND
  file_version_memberships.file_resource_id IN (
    SELECT id
    FROM file_resources
    WHERE file_data->'metadata'->>'mime_type' LIKE '%pdf%'
    OR file_data->'metadata'->>'mime_type' LIKE '%text%'
  )
;
```

Comes in at  `Time: 33.869 ms`

Reply:
Thanks for benchmarking! ⏱️

Reply:
Numbers don't lie... unless your math is wrong.